include/bm/bm_sim/actions.h

```diff
@@ -96,6 +97,42 @@ class NamedCalculation;
 class MeterArray;
 class CounterArray;
 
+// Internal representation of actions which are executed
+// periodically at a given interval
+struct PeriodicAction {
+  const char *name;
+  std::function<void()> fn;
+  std::chrono::milliseconds interval_ms;
+  unsigned int interval_ticks;
+};
+
+// Contains the list of the periodically executed actions
+class PeriodicActionsList {
```

Comment:
This class is not thread safe. Have you given some thought to what happens in case of P4 program swap?

Reply:
I'm unsure how the program swap occurs, and why it would cause concurrent access of this structure. Could you elaborate a bit on what causes this class to be less safe than ActionsOpcodesMap, or point me to documentation on the program swap procedure?

Reply:
Swapping the P4 program is done by loading a new JSON file (generated by the compiler). Unfortunately I do not believe there is any centralized documentation fore this, only comments scattered in the switch.h source code. It is very similar to the P4Runtime ForwardingPipelineConfig swap if you're familiar with P4Runtime. It is a 2 step operation:
* first, `load_new_config` is called (https://github.com/p4lang/behavioral-model/blob/master/include/bm/bm_sim/switch.h#L789) to load a new JSON (sent through RPC). All packet processing still uses the old JSON, but all subsequent runtime operations (e.g. table entry adds) are done on the new JSON
* when the application is ready, it calls `swap_configs` (https://github.com/p4lang/behavioral-model/blob/master/include/bm/bm_sim/switch.h#L792) which lets bmv2 know that the new JSON & associated runtime state are ready to be used for packet processing
There is actually a 3rd step. `swap_configs` only signals to bmv2 that configurations should be swapped. But the target (e.g. simple_switch) has final say and controls exactly when the swap occurs. This is because the target may have to update some internal state, e.g. pointers that may be invalidated by the swap. This is detailed here: https://github.com/p4lang/behavioral-model/blob/master/include/bm/bm_sim/switch.h#L49

Now back to your PR. `REGISTER_PERIODIC_ACTION` is actually executed when the first instance of a given primitive (extern function) is encountered while loading the JSON configuration. You have a single global instance of `PeriodicActionsList` with a thread periodically executing `next()`. What if `next()` is called while the new config is loaded and `REGISTER_PERIODIC_ACTION`? So that's why you need to make `PeriodicActionsList` thread safe.

When a new configuration is enabled, you need to cancel all the tasks for the previous P4 and start tasks for the new P4 IMO (the 2 programs may not be using the same extern functions). The only way to make that work is to tie the lifetime of `PeriodicActionsList` with that of `P4Objects`, by letting the `P4Objects` instance own the `PeriodicActionsList` instance just like it owns the `LearnEngine` instance. When configs are swapped, the `PeriodicActionsList` instances will be swapped as well. Now there is one issue with this approach: what about the periodic tasks registered outside of primitive / extern objects?

That leads us to what I think is the right approach. Let a `PeriodicActionsList` instance manage its own thread (like `LearnEngine`). Keep using the singleton pattern; even though it would make more sense to give ownership to `SwitchWContexts` IMO, it may be unpractical to achieve. But modify the `REGISTER_PERIODIC_ACTION` mechanics. Maybe it could look a lot like starting a thread actually.
```
// should not be copyable
// this is merely a skeleton
struct PeriodicTask {
    PeriodicTask(...) { PeriodicActionsList::get_instance()->register(...); }
    ~PeriodicTask() { cancel(); }
    void cancel() { PeriodicActionsList::get_instance()->unregister(...); }
};

class timeout_extern : public ActionPrimitive<> {
    timeout_extern() {
        // the task will be unregistered when the object is destroyed
        task = PeriodicTask(periodic, interval, this);
     }

    void operator() () {
        BMLOG_DEBUG("Calling timeout_extern");
    }

    // it is not a good idea to make this static, as it is not a good idea to have any shared static state in primitive instances
    // static void periodic() {
    void periodic() {
        BMLOG_DEBUG("This will be called 4 times a second if timeout_extern() is used in P4");
    }

    PeriodicTask task;
};
```
For a global static task, you could do `new PeriodicTask(...)` and never destroy it, or you could give ownership to the target class (e.g. `SimpleSwitch`).

This is still more convenient and efficient than having to manage a thread yourself for every periodic task you want to execute. You will also not have to worry about P4 program swap, as its `P4Objects` instance should have its own instance of each primitive in use.

Reply:
If you do agree to go this route, I believe you should put the new classes in their own source files, not in actions.h / actions.cpp